resources/knowledgebase/tla-review-guidelines.md

```diff
@@ -0,0 +1,147 @@
+
+* Check all defined invariants separately in the TLC configuration, not combined with other invariants as conjunction of invariants. This enables TLC to report which invariant is violated. See `tla-no-conjunct-of-invariants.md` for details.
+
+### Properties
+
+* Important properties should be declared as THEOREMs in the specification (e.g., `THEOREM Spec => EventuallyConsumed`). This serves as documentation of the key properties the spec is intended to satisfy and makes it clear what should be verified. However, don't include theorems that are not proven or checked with TLAPS or -at a minimum- checked with TLC.
+
+* Liveness properties should typically ensure repeated behavior rather than just single occurrences. For example, instead of checking that a consumer eventually consumes once (`<>(consumed > 0)`), check that consumers keep making progress (e.g., `[]<>(buffer = <<>>)` — the buffer is empty infinitely often). This ensures the system doesn't just satisfy the property once and then deadlock or stop making progress.
+
+* If liveness properties hold even without fairness, the property or fairness constraints need revision.
+
+### Readability and Documentation
+
+* Use meaningful and descriptive names for all identifiers (variables, constants, operators, actions). Names should clearly convey the purpose and meaning of the entity they represent. For example, prefer `buffer`, `Produce`, and `Consume` over generic names like `data`, `Action1`, and `Action2`.
+
+* Organize the specification logically with related definitions grouped together. Prioritize readability over space efficiency - specifications are meant to be understood by humans, not optimized for memory usage.
+
+* Each specification should include sufficient inline comments to explain the problem being modeled, key design decisions, and non-obvious logic. Comments should provide context that helps readers understand the specification without much external references. Consider using section comments to separate major parts of the specification (e.g., "---- Variables ----", "---- Actions ----").
+
+* Programmers have internalized the DRY principle. However, specs are short and rarely span multiple files. Thus, it is fine to relax the DRY principle. Specifically, specs tend to be easier to read if they are written in a more verbose style, and similar patterns are not extracted into common definitions.
+
+## Spec validation and verification with TLC
+
+### Configuration file
+
+* Consider whether deadlock checking is appropriate for your spec. Some specs legitimately terminate (e.g., when producers stop producing and consumers finish consuming all items from the buffer), and reporting this as a deadlock would be a false positive. In such cases, disable deadlock checking in the TLC configuration with `CHECK_DEADLOCK FALSE`.
```

Comment:
If you decide to disable deadlock checking, consider adding an invariant that characterizes states in which the system may terminate, for example
``(~ ENABLED Next) <=> buffer = <<>> /\ produced = DataItems``